_posts/2022-07-22-accessing-host-devices-in-flotta.markdown

```diff
@@ -0,0 +1,322 @@
+```
+
+The easiest way to identify the SELinux missing rules is using `audit2allow`. This utility analyzes the information in `/var/log/audit/audit.log` and generates the SELinux policies based on denied operations. In my case, I had to run the workload a few times until I was able to get all the rules.
+
+The utility created the following rules:
+
+```shell 
+module v4linux 1.0;
+
+require {
+	type v4l_device_t;
+	type container_t;
+	class chr_file { getattr ioctl map open read write };
+}
+
+#============= container_t ==============
+
+#!!!! This avc can be allowed using the boolean 'container_use_devices'
+allow container_t v4l_device_t:chr_file { getattr ioctl open read write };
+allow container_t v4l_device_t:chr_file map;
+```
+
+To create the package for the rules rules you can either use the `audit2allow -a -M <modulename>.pp` command, or you use the following commands for a step by step execution:
+
+Store the contents of the rules in a file named `v4linux.te` in the device's `/tmp`
+
+```shell
+[root@fedora ~]# cat >/tmp/webcam.te <<EOF
+module v4linux 1.0;
+
+require {
+	type v4l_device_t;
+	type container_t;
+	class chr_file { getattr ioctl map open read write };
+}
+
+#============= container_t ==============
+
+#!!!! This avc can be allowed using the boolean 'container_use_devices'
+allow container_t v4l_device_t:chr_file { getattr ioctl open read write };
+allow container_t v4l_device_t:chr_file map;
+EOF
+```
+
+Now convert it to a policy module:
+
+```shell
+[root@fedora ~]# checkmodule -M -m /tmp/v4linux.te -o /tmp/v4linux.mod
+```
+
+And generate the package:
+
+```shell
+[root@fedora ~]#  semodule_package -o /tmp/v4linux.pp -m /tmp/v4linux.mod
+```
+
+Finally install it:
+
+```shell
+[root@fedora ~]# semodule -i /tmp/v4linux.pp
+```
+
+Now SELinux is configured to allow the container to interact with the webcam. To test this, I deploy a new workload that does not have the `seLinuxOptions` defined in the pod spec:
+
+```yaml
+apiVersion: management.project-flotta.io/v1alpha1
+kind: EdgeWorkload
+metadata:
+  name: webcam-nosecuritycontext
+  annotations:
+    podman/run.oci.keep_original_groups: "1"
+spec:
+  deviceSelector:
+    matchLabels:
+      app: webcam
+  type: pod
+  pod:
+    spec:
+      containers:
+      - image: quay.io/jordigilh/ffmpeg:latest
+        name: fedora
+        volumeMounts:
+        - mountPath: /dev/video2
+          name: video
+      restartPolicy: Always      
+      volumes:
+      - name: video 
+        hostPath:
+          path: /dev/video2
+          type: File
+```
+
+Checking the status of the workload we get a satisfactory `Running`:
+
+```shell
+[jgil@fedora ~]$ kubectl get edgedevice/4cc27d11334f4242baa7efc795d2bbc0 -ojsonpath="{.status.workloads}"| jq .
+[
+  {
+    "lastTransitionTime": "2022-07-22T15:02:28Z",
+    "name": "webcam",
+    "phase": "Running"
+  },
+  {
+    "lastTransitionTime": "2022-07-22T15:31:28Z",
+    "name": "webcam-nosecuritycontext",
+    "phase": "Running"
```

Comment:
I'm surprised both workloads could run at the same time using the same device.

Reply:
well the first one was done with capturing the screenshots, so I guess there was no race condition in attempting to read/write from the device.